File: typescript/chapters/ch19/src/features/mailbox.ts
// mailbox 领域模型：定义持久消息、状态目录与 Store 协议，供 teammate 运行时和文件适配器共用；P16 在其上扩展结构化协议消息。
import { randomUUID } from "node:crypto";

import { z } from "zod";

import type { RuntimeEvent } from "../core/events.js";
import { isWindowsReservedComponent } from "../core/filesystem.js";

// 消息类型决定消费者如何解释 content，但不能改变 mailbox 的持久化协议。
export const MailboxMessageKind = Object.freeze({
  Task: "task",
  Message: "message",
  Result: "result",
});
export type MailboxMessageKind = (typeof MailboxMessageKind)[keyof typeof MailboxMessageKind];

// 协议消息在普通 mailbox 上新增 typed 扩展；request 与 response 的 kind 决定 approved 的取值。
export const ProtocolMessageKind = Object.freeze({
  ShutdownRequest: "shutdown_request",
  ShutdownResponse: "shutdown_response",
  PlanApprovalRequest: "plan_approval_request",
  PlanApprovalResponse: "plan_approval_response",
});
export type ProtocolMessageKind = (typeof ProtocolMessageKind)[keyof typeof ProtocolMessageKind];

// 每个目录名对应一个持久状态；同一条消息只允许在 ready/processing/done/quarantine 之一。
export const MailboxState = Object.freeze({
  Ready: "ready",
  Processing: "processing",
  Done: "done",
  Quarantine: "quarantine",
});
export type MailboxState = (typeof MailboxState)[keyof typeof MailboxState];

// 协议或持久化错误携带稳定 errorCode，工具边界可把它转成结构化 ToolResult。
export class MailboxError extends Error {
  readonly errorCode: string;

  constructor(errorCode: string, message: string, options?: ErrorOptions) {
    super(message, options);
    this.name = "MailboxError";
    this.errorCode = errorCode;
  }
}

// 文件损坏、锁失败和状态迁移冲突统一映射到同一错误族，调用方无需猜测底层异常。
export class MailboxStorageError extends MailboxError {
  constructor(message: string, options?: ErrorOptions) {
    super("mailbox_storage_error", message, options);
    this.name = "MailboxStorageError";
  }
}

// 邮箱消息以稳定 id、发送者和接收者建模，持久层据此保证投递与确认可追踪。
export interface MailboxMessage extends RuntimeEvent {
  readonly id: string;
  readonly sender: string;
  readonly recipient: string;
  readonly kind: MailboxMessageKind;
  readonly content: string;
  readonly createdAtUtc: Date;
  readonly eventId: string;
  readonly idempotencyKey: string;
}

// 协议消息额外携带 requestId 和 approved；request 只接受 null，response 必须携带布尔决策。
export interface ProtocolMailboxMessage extends RuntimeEvent {
  readonly id: string;
  readonly sender: string;
  readonly recipient: string;
  readonly kind: ProtocolMessageKind;
  readonly content: string;
  readonly createdAtUtc: Date;
  readonly requestId: string;
  readonly approved: boolean | null;
  readonly eventId: string;
  readonly idempotencyKey: string;
}

export type MailboxItem = MailboxMessage | ProtocolMailboxMessage;

// Store 暴露发送、认领和确认的协议操作，具体文件持久化留在适配器。
export interface MailboxStore {
  send(
    sender: string,
    recipient: string,
    content: string,
    kind: MailboxMessageKind,
  ): Promise<MailboxMessage>;
  claim(recipient: string): Promise<MailboxItem | undefined>;
  ack(message: MailboxItem): Promise<boolean>;
  release(message: MailboxItem): Promise<boolean>;
  quarantine(message: MailboxItem): Promise<boolean>;
  recoverProcessing(recipient: string): Promise<number>;
}

// 只有支持协议消息的 store 才实现 sendProtocol；Runtime 在构造边界做类型判定。
export interface ProtocolMailboxStore extends MailboxStore {
  sendProtocol(
    sender: string,
    recipient: string,
    content: string,
    kind: ProtocolMessageKind,
    options: { readonly requestId: string; readonly approved: boolean | null },
  ): Promise<ProtocolMailboxMessage>;
}

const CANONICAL_UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/u;
const AGENT_NAME = /^[a-z0-9]+(?:-[a-z0-9]+)*$/u;

export const spawnTeammateInputSchema = z
  .object({
    name: z.string(),
    role: z.string(),
    prompt: z.string(),
  })
  .strict();
export type SpawnTeammateInput = z.infer<typeof spawnTeammateInputSchema>;

export const sendMessageInputSchema = z
  .object({
    to: z.string(),
    content: z.string(),
  })
  .strict();
export type SendMessageInput = z.infer<typeof sendMessageInputSchema>;

// Agent 名会被用作文件目录名，因此同时校验 slug 规则和 Windows 保留组件。
export function canonicalAgentName(value: string): string {
  if (typeof value !== "string" || !AGENT_NAME.test(value) || isWindowsReservedComponent(value)) {
    throw new Error("Invalid or unsafe Agent name; expected a safe lowercase slug");
  }
  return value;
}

export function canonicalMailboxMessageId(value: string): string {
  // UUID 既是消息主键也是事件 idempotency key，必须使用规范格式。
  if (typeof value !== "string" || !CANONICAL_UUID.test(value)) {
    throw new Error("Mailbox message id must be a canonical UUID");
  }
  return value;
}

// 构造边界统一校验 id、sender、recipient、kind、content 和时钟，返回不可变消息。
export function createMailboxMessage(input: {
  id: string;
  sender: string;
  recipient: string;
  kind: MailboxMessageKind;
  content: string;
  createdAtUtc: Date;
}): MailboxMessage {
  const id = canonicalMailboxMessageId(input.id);
  const sender = canonicalAgentName(input.sender);
  const recipient = canonicalAgentName(input.recipient);
  if (!Object.values(MailboxMessageKind).includes(input.kind)) {
    throw new TypeError("Mailbox message kind must be task, message, or result");
  }
  if (typeof input.content !== "string" || input.content.trim().length === 0) {
    throw new Error("Mailbox message content must not be empty");
  }
  if (!(input.createdAtUtc instanceof Date) || !Number.isFinite(input.createdAtUtc.valueOf())) {
    throw new Error("Mailbox clock value must be a valid UTC Date");
  }
  const createdAtUtc = new Date(input.createdAtUtc.valueOf());
  return Object.freeze({
    id,
    sender,
    recipient,
    kind: input.kind,
    content: input.content,
    createdAtUtc,
    eventId: id,
    idempotencyKey: id,
    toPayload: mailboxPayload,
  });
}

// 协议消息构造边界额外校验 requestId 与 approved 的匹配关系，request 和 response 不能互相伪装。
export function createProtocolMailboxMessage(input: {
  id: string;
  sender: string;
  recipient: string;
  kind: ProtocolMessageKind;
  content: string;
  createdAtUtc: Date;
  requestId: string;
  approved: boolean | null;
}): ProtocolMailboxMessage {
  const id = canonicalMailboxMessageId(input.id);
  const sender = canonicalAgentName(input.sender);
  const recipient = canonicalAgentName(input.recipient);
  const requestId = canonicalMailboxMessageId(input.requestId);
  if (!Object.values(ProtocolMessageKind).includes(input.kind)) {
    throw new TypeError("ProtocolMailboxMessage kind must be a protocol kind");
  }
  if (typeof input.content !== "string" || input.content.trim().length === 0) {
    throw new Error("Mailbox message content must not be empty");
  }
  if (!(input.createdAtUtc instanceof Date) || !Number.isFinite(input.createdAtUtc.valueOf())) {
    throw new Error("Mailbox clock value must be a valid UTC Date");
  }
  const isResponse =
    input.kind === ProtocolMessageKind.ShutdownResponse ||
    input.kind === ProtocolMessageKind.PlanApprovalResponse;
  if (
    (isResponse && typeof input.approved !== "boolean") ||
    (!isResponse && input.approved !== null)
  ) {
    throw new TypeError(
      "Protocol response approved must be a boolean and request approved must be null",
    );
  }
  const createdAtUtc = new Date(input.createdAtUtc.valueOf());
  return Object.freeze({
    id,
    sender,
    recipient,
    kind: input.kind,
    content: input.content,
    createdAtUtc,
    requestId,
    approved: input.approved,
    eventId: id,
    idempotencyKey: id,
    toPayload: protocolMailboxPayload,
  });
}

// 磁盘 JSON 是外部输入，必须拒绝未知字段和非法值，避免坏消息污染运行态。
export function mailboxMessageFromJson(value: unknown): MailboxMessage {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new MailboxStorageError("Mailbox message payload is invalid");
  }
  const record = value as Record<string, unknown>;
  const expected = ["content", "created_at_utc", "id", "kind", "recipient", "sender"];
  const keys = Object.keys(record).sort();
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    throw new MailboxStorageError("Mailbox message payload contains unsupported fields");
  }
  if (
    typeof record.id !== "string" ||
    typeof record.sender !== "string" ||
    typeof record.recipient !== "string" ||
    typeof record.kind !== "string" ||
    typeof record.content !== "string" ||
    typeof record.created_at_utc !== "string"
  ) {
    throw new MailboxStorageError("Mailbox message payload is invalid");
  }
  if (!Object.values(MailboxMessageKind).includes(record.kind as MailboxMessageKind)) {
    throw new MailboxStorageError("Mailbox message kind is invalid");
  }
  const createdAtUtc = new Date(record.created_at_utc);
  if (
    !record.created_at_utc.endsWith("Z") ||
    !Number.isFinite(createdAtUtc.valueOf()) ||
    createdAtUtc.toISOString() !== record.created_at_utc
  ) {
    throw new MailboxStorageError("Mailbox message timestamp is invalid");
  }
  try {
    return createMailboxMessage({
      id: record.id,
      sender: record.sender,
      recipient: record.recipient,
      kind: record.kind as MailboxMessageKind,
      content: record.content,
      createdAtUtc,
    });
  } catch (error) {
    throw new MailboxStorageError("Mailbox message fields failed validation", { cause: error });
  }
}

// 普通与协议消息共用反序列化入口；先按 kind 分流，再走严格字段校验。
export function mailboxItemFromJson(value: unknown): MailboxItem {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new MailboxStorageError("Mailbox message payload is invalid");
  }
  const record = value as Record<string, unknown>;
  if (typeof record.kind !== "string")
    throw new MailboxStorageError("Mailbox message kind is invalid");
  if (Object.values(MailboxMessageKind).includes(record.kind as MailboxMessageKind)) {
    return mailboxMessageFromJson(record);
  }
  const expected = [
    "approved",
    "content",
    "created_at_utc",
    "id",
    "kind",
    "recipient",
    "request_id",
    "sender",
  ];
  const keys = Object.keys(record).sort();
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    throw new MailboxStorageError("Protocol mailbox payload contains unsupported fields");
  }
  if (
    typeof record.id !== "string" ||
    typeof record.sender !== "string" ||
    typeof record.recipient !== "string" ||
    typeof record.content !== "string" ||
    typeof record.created_at_utc !== "string" ||
    typeof record.request_id !== "string" ||
    (record.approved !== null && typeof record.approved !== "boolean")
  ) {
    throw new MailboxStorageError("Protocol mailbox payload is invalid");
  }
  const createdAtUtc = new Date(record.created_at_utc);
  if (
    !record.created_at_utc.endsWith("Z") ||
    !Number.isFinite(createdAtUtc.valueOf()) ||
    createdAtUtc.toISOString() !== record.created_at_utc
  ) {
    throw new MailboxStorageError("Mailbox message timestamp is invalid");
  }
  try {
    return createProtocolMailboxMessage({
      id: record.id,
      sender: record.sender,
      recipient: record.recipient,
      kind: record.kind as ProtocolMessageKind,
      content: record.content,
      createdAtUtc,
      requestId: record.request_id,
      approved: record.approved,
    });
  } catch (error) {
    throw new MailboxStorageError("Protocol mailbox fields failed validation", { cause: error });
  }
}

// 持久化使用 snake_case 的稳定字段名，内存对象保持 camelCase，转换只在边界发生。
export function mailboxMessageToJson(message: MailboxMessage): Readonly<Record<string, string>> {
  return Object.freeze({
    id: message.id,
    sender: message.sender,
    recipient: message.recipient,
    kind: message.kind,
    content: message.content,
    created_at_utc: message.createdAtUtc.toISOString(),
  });
}

// 普通与协议消息共用序列化入口；协议字段只在边界出现，避免污染普通消息。
export function mailboxItemToJson(
  message: MailboxItem,
): Readonly<Record<string, string | boolean | null>> {
  if (isProtocolMailboxMessage(message)) {
    return Object.freeze({
      id: message.id,
      sender: message.sender,
      recipient: message.recipient,
      kind: message.kind,
      content: message.content,
      created_at_utc: message.createdAtUtc.toISOString(),
      request_id: message.requestId,
      approved: message.approved,
    });
  }
  return mailboxMessageToJson(message);
}

export function equalMailboxMessages(left: MailboxMessage, right: MailboxMessage): boolean {
  // 幂等 ack 需要按完整消息内容比较，不能只比较 id，否则可能确认了错误的负载。
  return (
    left.id === right.id &&
    left.sender === right.sender &&
    left.recipient === right.recipient &&
    left.kind === right.kind &&
    left.content === right.content &&
    left.createdAtUtc.valueOf() === right.createdAtUtc.valueOf()
  );
}

export function equalMailboxItems(left: MailboxItem, right: MailboxItem): boolean {
  const leftProtocol = isProtocolMailboxMessage(left);
  const rightProtocol = isProtocolMailboxMessage(right);
  if (leftProtocol !== rightProtocol) return false;
  if (!leftProtocol && !rightProtocol) {
    return equalMailboxMessages(left, right);
  }
  const leftProtocolMessage = left as ProtocolMailboxMessage;
  const rightProtocolMessage = right as ProtocolMailboxMessage;
  return (
    leftProtocolMessage.id === rightProtocolMessage.id &&
    leftProtocolMessage.sender === rightProtocolMessage.sender &&
    leftProtocolMessage.recipient === rightProtocolMessage.recipient &&
    leftProtocolMessage.kind === rightProtocolMessage.kind &&
    leftProtocolMessage.content === rightProtocolMessage.content &&
    leftProtocolMessage.createdAtUtc.valueOf() === rightProtocolMessage.createdAtUtc.valueOf() &&
    leftProtocolMessage.requestId === rightProtocolMessage.requestId &&
    leftProtocolMessage.approved === rightProtocolMessage.approved
  );
}

// 通过 payload 的稳定标记识别协议消息，不依赖字符串前缀或鸭子类型字段。
export function isProtocolMailboxMessage(message: RuntimeEvent): message is ProtocolMailboxMessage {
  const payload = message.toPayload();
  return payload.kind === "protocol" && "request_id" in payload && "protocol_kind" in payload;
}

// 构造期用 sendProtocol 能力区分普通 MailboxStore 与协议扩展 store。
export function isProtocolMailboxStore(store: MailboxStore): store is ProtocolMailboxStore {
  return typeof Reflect.get(store, "sendProtocol") === "function";
}

export function randomMailboxMessageId(): string {
  // 默认 id 生成器只在生产构造时使用；测试可注入确定 UUID。
  return randomUUID();
}

// RuntimeEvent 的 payload 暴露结构化字段，避免调用方通过字符串前缀判断消息类型。
function mailboxPayload(this: MailboxMessage): Readonly<Record<string, unknown>> {
  return Object.freeze({
    kind: "mailbox",
    message_id: this.id,
    sender: this.sender,
    recipient: this.recipient,
    message_kind: this.kind,
    content: this.content,
    created_at_utc: this.createdAtUtc.toISOString(),
  });
}

// 协议 payload 额外携带 request_id 与 approved，供 EventInbox 和协议路由区分消息类型。
function protocolMailboxPayload(this: ProtocolMailboxMessage): Readonly<Record<string, unknown>> {
  return Object.freeze({
    kind: "protocol",
    message_id: this.id,
    sender: this.sender,
    recipient: this.recipient,
    protocol_kind: this.kind,
    content: this.content,
    created_at_utc: this.createdAtUtc.toISOString(),
    request_id: this.requestId,
    approved: this.approved,
  });
}

File: python/ch16_agent/agent_ch16/features/mailbox.py
"""Mailbox 领域模型。

Java 类比：``MailboxMessage`` 是不可变 record，``MailboxStore`` 是 Repository 接口。
本文件只定义消息规则和状态机，不关心消息最终使用文件、数据库还是消息队列保存。

第 16 章关键：ProtocolMailboxMessage 是 typed 协议消息，携带 request_id 和 approved 字段，
模型不需要解析自然语言就能识别协议类型和审批结果。
"""

from __future__ import annotations

import datetime as dt
import re
import uuid
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Literal, Protocol

MailboxMessageKind = Literal["task", "message", "result"]  # 普通消息三种类型
ProtocolMessageKind = Literal[
    "shutdown_request", "shutdown_response", "plan_approval_request", "plan_approval_response"
]  # 结构化协议消息四种类型（请求/响应 × 关机/计划审批）
MailboxState = Literal["ready", "processing", "done", "quarantine"]  # 消息四态
MAILBOX_KINDS: tuple[MailboxMessageKind, ...] = ("task", "message", "result")
MAILBOX_STATES: tuple[MailboxState, ...] = ("ready", "processing", "done", "quarantine")
PROTOCOL_KINDS: tuple[ProtocolMessageKind, ...] = (
    "shutdown_request",
    "shutdown_response",
    "plan_approval_request",
    "plan_approval_response",
)
LEAD_NAME = "lead"  # Lead Agent 的固定身份名

_UUID = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
_AGENT_NAME = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")  # 安全的小写 slug
_WINDOWS_RESERVED = {
    "con",
    "prn",
    "aux",
    "nul",
    *(f"com{i}" for i in range(1, 10)),
    *(f"lpt{i}" for i in range(1, 10)),
}  # Windows 保留设备名，不能用作目录名


class MailboxError(Exception):
    """Mailbox 领域异常，``error_code`` 可用于工具的机器可读错误码。

    这是什么：Mailbox 系统的基础异常类
    Java 类比：类似自定义的 BusinessException，error_code 相当于枚举错误码
    为什么需要：让调用方能根据 error_code 精确识别失败原因并做针对性处理
    """

    def __init__(self, error_code: str, message: str) -> None:
        super().__init__(message)
        self.error_code = error_code  # 机器可读的错误码


class MailboxStorageError(MailboxError):
    """文件损坏、状态冲突或原子迁移失败。

    这是什么：Mailbox 持久化层的异常
    Java 类比：类似 DataAccessException
    为什么需要：区分领域错误和存储错误，存储错误通常需要重试或人工介入
    """

    def __init__(self, message: str) -> None:
        super().__init__("mailbox_storage_error", message)


def canonical_agent_name(value: str) -> str:
    """校验可安全用作目录名的 Agent 小写 slug。

    这是什么：规范化 Agent 身份名，确保可以安全用作目录名
    Java 类比：类似输入校验的静态工具方法
    为什么需要：防止目录穿越攻击和 Windows 保留设备名冲突
    """
    if (
        not isinstance(value, str)
        or _AGENT_NAME.fullmatch(value) is None
        or value.lower() in _WINDOWS_RESERVED
    ):
        raise ValueError("Agent 名必须是安全的小写 slug，例如 alice 或 api-writer")
    return value


def canonical_message_id(value: str) -> str:
    """校验 canonical UUID；它同时充当事件 ID 和幂等键。

    这是什么：校验消息 ID 格式，必须是标准 UUID
    Java 类比：类似 UUID.fromString() 的校验逻辑
    为什么需要：消息 ID 同时用于去重和幂等，必须保证全局唯一性和格式一致性
    """
    if not isinstance(value, str) or _UUID.fullmatch(value) is None:
        raise ValueError("Mailbox 消息 id 必须是 canonical UUID")
    return value


@dataclass(frozen=True, slots=True)
class MailboxMessage:
    """一条可持久化、可作为 RuntimeEvent 注入 Agent Loop 的消息。

    这是什么：普通 Mailbox 消息，不可变 record
    Java 类比：类似不可变 record MailboxMessage(id, sender, recipient, kind, content, createdAt)
    为什么需要：作为队友间通信的基础消息载体，支持持久化和事件驱动
    """

    id: str  # 全局唯一消息主键（UUID），也是事件去重 ID 和幂等键
    sender: str  # 发送方 Agent 身份，例如 lead、alice
    recipient: str  # 接收方 Agent 身份，同时决定邮箱目录（.agent_tutorial/mailbox/{recipient}/）
    kind: MailboxMessageKind  # task（任务分配）、message（普通消息）或 result（结果反馈）
    content: str  # 原始正文；保存时不自动 strip，避免改变业务文本
    created_at_utc: dt.datetime  # 带 UTC 时区的创建时间，用于 FIFO 排序

    def __post_init__(self) -> None:
        canonical_message_id(self.id)
        canonical_agent_name(self.sender)
        canonical_agent_name(self.recipient)
        if self.kind not in MAILBOX_KINDS:
            raise ValueError("Mailbox 消息 kind 必须是 task、message 或 result")
        if not isinstance(self.content, str) or not self.content.strip():
            raise ValueError("Mailbox 消息正文不能为空")
        if self.created_at_utc.tzinfo is None or self.created_at_utc.utcoffset() != dt.timedelta(0):
            raise ValueError("created_at_utc 必须是带 UTC 时区的时间")

    @property
    def event_id(self) -> str:
        """RuntimeEvent 的稳定去重 ID。"""
        return self.id

    @property
    def context_identity(self) -> str:
        """事件回合使用接收方身份，Lead 消息因此以 lead 执行。"""
        return self.recipient

    @property
    def idempotency_key(self) -> str:
        """工具副作用使用消息 UUID 去重。"""
        return self.id

    @property
    def prompt(self) -> str:
        """队友或 Lead 真正交给模型处理的正文。"""
        return self.content

    def to_payload(self) -> Mapping[str, object]:
        """转换为事件消息中的稳定 JSON 字段。"""
        return {
            "kind": "mailbox",
            "message_id": self.id,
            "sender": self.sender,
            "recipient": self.recipient,
            "message_kind": self.kind,
            "content": self.content,
            "created_at_utc": format_utc(self.created_at_utc),
        }


@dataclass(frozen=True, slots=True)
class ProtocolMailboxMessage:
    """带 request_id/approved 的结构化协议消息。

    这是什么：类型化的协议消息，携带结构化字段用于计划审批和优雅关机
    Java 类比：类似带领域字段的事件 record，ProtocolMailboxMessage extends MailboxMessage
    为什么需要：让模型不需要解析自然语言就能识别协议类型和审批结果，避免理解歧义

    第 16 章核心：
    - request_id 关联 ProtocolRequest，保证消息与持久化请求的一致性
    - approved 字段是机器可读的审批结论（True=批准，False=拒绝）
    - kind 区分请求/响应和关机/计划审批四种类型
    """

    id: str  # 当前传输消息 UUID（每次传输生成新 ID）
    sender: str  # 发送方身份（请求：队友，响应：lead / 关机请求：lead，响应：队友）
    recipient: str  # 接收方身份（请求：lead，响应：队友 / 关机请求：队友，响应：lead）
    kind: ProtocolMessageKind  # shutdown/plan_approval 的 request/response 四种类型
    content: str  # 展示给对方的正文或反馈（人类可读）
    created_at_utc: dt.datetime  # 传输创建时间（UTC）
    request_id: str  # 关联 ProtocolRequest 的 UUID（同一协议请求的请求和响应共享此 ID）
    approved: bool | None  # 请求必须为 None，响应必须为 bool（True=批准，False=拒绝）

    def __post_init__(self) -> None:
        canonical_message_id(self.id)
        canonical_message_id(self.request_id)
        canonical_agent_name(self.sender)
        canonical_agent_name(self.recipient)
        if self.kind not in PROTOCOL_KINDS:
            raise ValueError("协议消息 kind 无效")
        if not isinstance(self.content, str) or not self.content.strip():
            raise ValueError("协议消息正文不能为空")
        if self.created_at_utc.tzinfo is None or self.created_at_utc.utcoffset() != dt.timedelta(0):
            raise ValueError("created_at_utc 必须是 UTC 时间")
        response = self.kind.endswith("_response")
        if response != isinstance(self.approved, bool):
            raise ValueError("协议响应 approved 必须是 bool，请求必须是 None")

    @property
    def event_id(self) -> str:
        """RuntimeEvent 去重 ID。"""
        return self.id

    @property
    def context_identity(self) -> str:
        """事件回合使用收件方身份。"""
        return self.recipient

    @property
    def idempotency_key(self) -> str:
        """协议副作用幂等键。"""
        return self.id

    @property
    def prompt(self) -> str:
        """协议路由后的模型提示由运行时另行生成。"""
        return self.content

    def to_payload(self) -> Mapping[str, object]:
        """转换为带协议标记的事件字段。"""
        return {
            "kind": "protocol",
            "message_id": self.id,
            "sender": self.sender,
            "recipient": self.recipient,
            "protocol_kind": self.kind,
            "content": self.content,
            "created_at_utc": format_utc(self.created_at_utc),
            "request_id": self.request_id,
            "approved": self.approved,
        }


MailboxItem = MailboxMessage | ProtocolMailboxMessage


class MailboxStore(Protocol):
    """Mailbox Repository 协议；状态迁移由实现保证原子性。

    这是什么：Mailbox 的持久化接口，定义消息的 CRUD 和状态迁移
    Java 类比：类似 Repository<MailboxMessage> 接口
    为什么需要：抽象持久化层，让领域逻辑不依赖具体存储实现（文件/数据库/消息队列）

    四态状态机：
    - ready: 新消息，等待被领取
    - processing: 已被 claim，正在处理
    - done: 已被 ack，处理完成
    - quarantine: 处理失败，需要人工介入
    """

    def send(
        self, sender: str, recipient: str, content: str, kind: MailboxMessageKind
    ) -> MailboxMessage:
        """创建并保存新消息到 ready 状态。"""
        ...
    def claim(self, recipient: str) -> MailboxItem | None:
        """原子领取一条 ready 消息，迁移到 processing 状态。"""
        ...
    def ack(self, message: MailboxItem) -> bool:
        """确认消息已处理完成，从 processing 迁移到 done。"""
        ...
    def release(self, message: MailboxItem) -> bool:
        """释放消息租约，从 processing 迁移回 ready（处理失败时使用）。"""
        ...
    def quarantine(self, message: MailboxItem) -> bool:
        """标记消息为隔离状态，从 processing 迁移到 quarantine（需要人工介入）。"""
        ...
    def recover_processing(self, recipient: str) -> int:
        """恢复僵尸 processing 消息到 ready（进程崩溃后使用）。"""
        ...


def new_message(
    sender: str,
    recipient: str,
    content: str,
    kind: MailboxMessageKind,
    *,
    message_id: str | None = None,
    now: dt.datetime | None = None,
) -> MailboxMessage:
    """创建通过统一领域校验的消息。"""
    timestamp = (now or dt.datetime.now(dt.UTC)).astimezone(dt.UTC)
    # JSON 协议只保存毫秒；构造时就截断微秒，保证内存对象和磁盘读回对象完全相等。
    timestamp = timestamp.replace(microsecond=(timestamp.microsecond // 1000) * 1000)
    return MailboxMessage(
        canonical_message_id(message_id or str(uuid.uuid4())),
        canonical_agent_name(sender),
        canonical_agent_name(recipient),
        kind,
        content,
        timestamp,
    )


def format_utc(value: dt.datetime) -> str:
    """输出严格带 ``Z`` 的 UTC ISO 字符串。"""
    return value.astimezone(dt.UTC).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def messages_equal(left: MailboxMessage, right: MailboxMessage) -> bool:
    """按完整快照比较，防止只凭相同 ID 错误确认不同正文。"""
    return left == right


def is_protocol_message(value: object) -> bool:
    """判断事件是否为 typed 协议消息。"""
    return isinstance(value, ProtocolMailboxMessage)


def protocol_messages_equal(left: ProtocolMailboxMessage, right: ProtocolMailboxMessage) -> bool:
    """按完整协议消息比较，防止错误 ack。"""
    return left == right
